### PR TITLE
feat: implement fri commitment

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements FRI commitment. Therefore this introduces new class `FRI`, which is another child class of `UnivariatePolynomialCommitmentScheme`. In addition, this fixes a bug in `BinaryMerkleTree` and make `Goldilocks` FFT friendly field.

 See here for references!
- [fri_commit_phase()](https://github.com/lambdaclass/lambdaworks/blob/703db8a535793ecc820e044b06c1b735764f0881/exercises/message/src/starks/fri/mod.rs#L20-L72),
- [fri_decommit_phase()](https://github.com/lambdaclass/lambdaworks/blob/703db8a535793ecc820e044b06c1b735764f0881/exercises/message/src/starks/fri/mod.rs#L74-L127) 
- [Anatomy of a STARK, Part 3: FRI](https://aszepieniec.github.io/stark-anatomy/fri.html)

### CODE REVIEW

File: tachyon/crypto/commitments/vector_commitment_scheme.h [hunk 1]
```diff
@@ -94,8 +94,10 @@ class VectorCommitmentScheme {
 
   // Verify an opening |proof| that proves that |members| belong to a
   // |commitment|.
+  // NOTE(chokobole): const was removed from |Commitment| since it can be a
+  // |Transcript|. At this moment, |WriteToTranscript()| is not a const method.
```
Comment (Insun35) [inline tachyon/crypto/commitments/vector_commitment_scheme.h:98]: Is there any case where `WriteToTranscript` is needed in verification?
Plus, wouldn't `proof` be more appropriate for variables for getting transcripts?

File: tachyon/crypto/commitments/fri/fri.h [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    size_t k = domain->log_size_of_group();
+    if (k > 1) {
+      sub_domains_ = base::CreateVector(k - 1, [k](size_t i) {
+        return Domain::Create(size_t{1} << (k - i - 1));
+      });
+    }
+    storage_->Allocate(k);
+  }
+
+  // UnivariatePolynomialCommitmentScheme methods
```
Comment (dongchangYoo) [inline tachyon/crypto/commitments/fri/fri.h:45]: This is a method of the `FRI`, and it appears that the base classes are using `FRI::N()`. Please confirm if the meaning of this comment is correct.

File: tachyon/crypto/commitments/vector_commitment_scheme.h (hunk 1 above)
Comment (chokobole) [reply tachyon/crypto/commitments/vector_commitment_scheme.h:98]: 1. `ReadFromProof` internally calls `WriteToTranscript`! 
2. I don't understand a question, I'll ask this at office tomorrow!

File: tachyon/crypto/commitments/fri/fri.h (hunk 2 above)
Comment (chokobole) [reply tachyon/crypto/commitments/fri/fri.h:45]: I follow this convention when I write some methods which are either or virtual methods or needed by parent class

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
Fix typo in [7dca00f](https://github.com/kroma-network/tachyon/pull/194/commits/7dca00f04cfddb3af842074c25d75bbd2c943125) please. UnivaraitePolynomial -> UnivariatePolynomial

File: tachyon/crypto/commitments/vector_commitment_scheme.h (hunk 1 above)
Comment (chokobole) [reply tachyon/crypto/commitments/vector_commitment_scheme.h:98]: `Commitment` sometimes can be a `Transcript`.

**Review by Insun35 — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@chokobole Is there a reason why you made the change in [26cb67e](https://github.com/kroma-network/tachyon/pull/194/commits/26cb67e75176bc49be9d6f22783d62a7b34177cb)? `SimpleHasher` and `SimpleFRIStorage` for fri test is in anonymous namespace, so for me, it seems not unified.

File: tachyon/crypto/commitments/fri/fri_proof.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Use of this source code is governed by a Apache-2.0 style license that
+// can be found in the LICENSE.lambdaworks.
+
+#ifndef TACHYON_CRYPTO_COMMITMENTS_FRI_FRI_PROOF_H_
+#define TACHYON_CRYPTO_COMMITMENTS_FRI_FRI_PROOF_H_
+
+#include <vector>
+
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_proof.h"
+#include "tachyon/crypto/transcripts/transcript.h"
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/fri/fri_proof.h:10]: Please remove this header.

File: tachyon/crypto/commitments/fri/fri.h [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        if (!tree.Commit(evals.evaluations(), &root)) return false;
+        if (!writer->WriteToProof(root)) return false;
+        cur_poly = &folded_poly;
+      }
+    }
+
+    beta = writer->SqueezeChallenge();
+    folded_poly = cur_poly->template Fold<false>(beta);
+    const F* constant = folded_poly[0];
+    root = constant == nullptr ? F::Zero() : *constant;
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/fri/fri.h:76]:
```suggestion
    root = constant ? F::Zero() : *constant;
```

File: tachyon/crypto/commitments/fri/fri.h [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+      //
+      // Next layer equation:
+      // Pᵢ₊₁(X) = Pᵢ_even(X²) + β * Pᵢ_odd(X²)
+      //
+      // If the domain of Pᵢ(X) is Dᵢ = {ω⁰, ω¹, ..., ωⁿ⁻¹},
+      // then the domain of Pᵢ₊₁(X) is Dᵢ₊₁ = {ω⁰, ω¹, ..., ωᵏ⁻¹},
+      // where k = n / 2.
+      //
+      // As per the definition:
+      // Pᵢ₊₁(ω²ʲ) = Pᵢ_even((ωʲ)²) + β * Pᵢ_odd((ωʲ)²)
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/fri/fri.h:151]:
```suggestion
      // Pᵢ₊₁(ωʲ) = Pᵢ_even((ωʲ)²) + β * Pᵢ_odd((ωʲ)²)
```

File: tachyon/crypto/commitments/fri/fri.h [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+      //
+      // Substituting Pᵢ_even and Pᵢ_odd:
+      // Pᵢ₊₁(ω²ʲ) = (Pᵢ(ωʲ) + Pᵢ(-ωʲ)) / 2 + β * (Pᵢ(ωʲ) - Pᵢ(-ωʲ)) / (2 * ωʲ)
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/fri/fri.h:154]:
```suggestion
      // Pᵢ₊₁(ωʲ) = (Pᵢ(ωʲ) + Pᵢ(-ωʲ)) / 2 + β * (Pᵢ(ωʲ) - Pᵢ(-ωʲ)) / (2 * ωʲ)
```

File: tachyon/crypto/commitments/fri/fri.h [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        evaluation_sym = proof.evaluations_sym[i];
+        x = domain_->GetElement(leaf_index);
+      } else {
+        evaluation *= (F::One() + beta);
+        evaluation_sym *= (F::One() - beta);
+        evaluation += evaluation_sym;
+        evaluation *= two_inv;
+
+        if (evaluation != proof.evaluations[i]) {
+          LOG(ERROR)
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/fri/fri.h:168]: Can you add #include "tachyon/base/logging.h"?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/crypto/commitments/fri/fri.h [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+template <typename F, size_t MaxDegree>
+class FRI : public UnivariatePolynomialCommitmentScheme<FRI<F, MaxDegree>> {
+ public:
+  using Base = UnivariatePolynomialCommitmentScheme<FRI<F, MaxDegree>>;
+  using Poly = typename Base::Poly;
+  using Evals = typename Base::Evals;
+  using Domain = typename Base::Domain;
+
+  FRI() = default;
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/crypto/commitments/fri/fri.h:28]: It depends on the case, and at this moment, it is needed by `fri_unittest.cc`!

**Comment by chokobole** [comment]
> @chokobole Is there a reason why you made the change in [26cb67e](https://github.com/kroma-network/tachyon/pull/194/commits/26cb67e75176bc49be9d6f22783d62a7b34177cb)? `SimpleHasher` and `SimpleFRIStorage` for fri test is in anonymous namespace, so for me, it seems not unified.

Because I wanted `SimpleBinaryMerkleTreeStorage` to be shared between `fri_unittest.cc` and `binary_merkle_tree_unittest.cc`

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**